[NFC][Sema] Move `Sema::AssignmentAction` into its own scoped enum

delcypher opened the pull request:
The primary motivation behind this is to allow the enum type to be
referred to earlier in the Sema.h file which is needed for #106321.

It was requested in #106321 that a scoped enum be used (rather than
moving the enum declaration earlier in the Sema class declaration).
Unfortunately doing this creates a lot of churn as all use sites of the
enum constants had to be changed. Appologies to all downstream forks in
advanced.

Note the AA_ prefix has been dropped from the enum value names as they
are now redundant.

delcypher commented:
@Endilll This is my attempt at implementing what you asked for in https://github.com/llvm/llvm-project/pull/106321#discussion_r1734379624

I'm not a huge fan of this as I think this will be extremely annoying for downstream forks with very little upside vs just moving the unscoped enum to be earlier in the `Sema` class declaration (which creates no churn).

However, if you and @AaronBallman want this version of the PR then we can with this.

Endilll commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -204,6 +204,24 @@ class SemaPPCallbacks;
 
+// AssignmentAction - This is used by all the assignment diagnostic functions
+// to represent what is actually causing the operation
+enum class AssignmentAction : unsigned {
```
```suggestion
enum class AssignmentAction {
```

Endilll commented:
> @Endilll This is my attempt at implementing what you asked for in [#106321 (comment)](https://github.com/llvm/llvm-project/pull/106321#discussion_r1734379624)
> 
> I'm not a huge fan of this as I think this will be extremely annoying for downstream forks with very little upside vs just moving the unscoped enum to be earlier in the `Sema` class declaration (which creates no churn).
> 
> However, if you and @AaronBallman want this version of the PR then we can go with this.

Yeah, no worries, we've been doing this on-demand for some time now. The biggest downside of unscoped enums, especially nested in classes, is that they are not eligible to be forward declared without including the header they are defined in. So you're following a well-established precedent here.

Endilll commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -204,6 +204,24 @@ class SemaPPCallbacks;
+};
+inline const StreamingDiagnostic &operator<<(const StreamingDiagnostic &DB,
+                                             const AssignmentAction &AA) {
+  DB << (unsigned)AA;
```
I backported `std::to_underlying` as `llvm::to_underlying` for this exact purpose.

delcypher commented:
@Endilll I've tried to address your feedback.

Endilll reviewed (approved):
Thank you for resolving your issue the hard way, and paying some of the technical debt we accumulated over the years.

shafik commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -8819,8 +8822,8 @@ Expr *OpenMPIterationSpaceChecker::buildNumIterations(
-        Diff = SemaRef.PerformImplicitConversion(Diff.get(), NewType,
-                                                 Sema::AA_Converting, true);
+        Diff = SemaRef.PerformImplicitConversion(
+            Diff.get(), NewType, AssignmentAction::Converting, true);
```
```suggestion
            Diff.get(), NewType, AssignmentAction::Converting, /*AllowExplicit=*/true);
```

nitpick

shafik commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -9469,7 +9473,8 @@ static ExprResult buildCounterUpdate(
       Update = SemaRef.PerformImplicitConversion(
-          Update.get(), VarRef.get()->getType(), Sema::AA_Converting, true);
+          Update.get(), VarRef.get()->getType(), AssignmentAction::Converting,
+          true);
```
```suggestion
          /*AllowExplicit=*/true);
```

shafik commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -9491,8 +9496,8 @@ static ExprResult widenIterationCount(unsigned Bits, Expr *E, Sema &SemaRef) {
-  return SemaRef.PerformImplicitConversion(E, NewType, Sema::AA_Converting,
-                                           true);
+  return SemaRef.PerformImplicitConversion(E, NewType,
+                                           AssignmentAction::Converting, true);
```
```suggestion
                                           AssignmentAction::Converting, /*AllowExplicit=*/true);
```

shafik reviewed (approved):
LGTM except for nitpicks

delcypher committed:
[NFC][Sema] Move Sema::AssignmentAction into its own scoped enum

The primary motivation behind this is to allow the enum type to be
referred to earlier in the Sema.h file which is needed for #106321.

It was requested in #106321 that a scoped enum be used (rather than
moving the enum declaration earlier in the Sema class declaration).
Unfortunately doing this creates a lot of churn as all use sites of the
enum constants had to be changed. Appologies to all downstream forks in
advanced.

Note the AA_ prefix has been dropped from the enum value names as they
are now redundant.
